web/hooks/use-subscription.ts

```diff
@@ -0,0 +1,253 @@
+const callbacks: {
+  hookId: string
+  callback: (data: any) => void
+  filter: string
+  table: string
+}[] = []
+let mainChannelResubscribeInterval: NodeJS.Timer | undefined = undefined
+
+export function useSubscription<T extends TableName>(
+  table: T,
+  db: SupabaseClient,
+  filter?: RowFilter<T>,
+  getInitialValues?: () => Promise<RowFor<T>[]>
+) {
+  const [values, setValues] = useState<RowFor<T>[]>([])
+  const hookId = useId()
+  const primaryKeyColumns = primaryKeyColumnsByTable[table]
+  const valuesToDelete = useRef<RowFor<T>[]>([])
+  const retrievedInitialValues = useRef<boolean>(false)
+  const channelResubscribeInterval = useRef<NodeJS.Timer>()
+
+  const rowMatches = (a: RowFor<T>, b: RowFor<T>) => {
+    return primaryKeyColumns.every((key) => a[key] === b[key])
+  }
+
+  const getMyCallback = () => {
+    return {
+      hookId,
+      callback: updateValuesOnPostgresEvent,
+      filter: filter ? `${filter.k as string}=eq.${filter.v}` : '',
+      table,
+    }
+  }
+
+  useEffect(() => {
+    initChannelAndCallbacks()
+  }, [db])
+
+  const initChannelAndCallbacks = () => {
+    // very first channel opened
+    if (!mainChannel) subscribeToNewChannel()
+    // channel already exists, restart and resubscribe callbacks, adding my own
+    else restartChannel()
+  }
+
+  const restartChannel = async () => {
+    console.log('Restarting channel')
+    subscriptionStatus = 'CONNECTING'
+    // There may be updates while we've disconnected, so we need to refresh the values
+    retrievedInitialValues.current = false
+    if (mainChannel) {
+      await mainChannel.unsubscribe()
+      await db.removeChannel(mainChannel)
+      mainChannel = undefined
+    }
+    initChannelAndCallbacks()
+  }
+
+  // Must be called only once per channel
+  const subscribeToNewChannel = () => {
+    const channel = db.channel(CHANNEL_NAME)
+    callbacks.push(getMyCallback())
+    setChannelCallbacks(channel)
+    channel.subscribe(async (status, err) => {
+      console.log('channel subscribe status', status, err)
+      if (
+        status !== 'SUBSCRIBED' &&
+        subscriptionStatus !== 'CONNECTING' &&
+        !mainChannelResubscribeInterval
+      ) {
+        startResubscribeLoop()
+      }
+      subscriptionStatus = status
+    })
+    mainChannel = channel
+  }
+
+  const updateValuesOnPostgresEvent = (payload: any) => {
+    setValues((prev) => {
+      const { new: newRecord, old: oldRecord } = payload
+      // During DELETE operation, oldValue will have the primary key values set and NO data
+      const oldValue = oldRecord as RowFor<T>
+      const index = prev.findIndex((m) => rowMatches(m, oldValue))
+      // New row with filled values
+      if (newRecord && primaryKeyColumns.every((k) => newRecord[k] !== null)) {
+        const newValue = newRecord as RowFor<T>
+        // We already have this row, so update it
+        if (oldRecord && index > -1) {
+          prev[index] = newValue
+          return prev
```

Comment:
Does it work to mutate the old array? I thought setState would not trigger a re-render if the reference is the same.

Reply:
Oh it looks like creating a new array is more robust, will do that instead